feat: support logout in the pre-exit

v1v opened the pull request:
#### What

- Add support for logout with a new argument to disable the logout if necessary.
- Add local development
- Add UTs
- Use local development Make goals in the CI too.

#### Why

Avoid leftovers when using docker login or similar CLI tools.
Unify the DX for local development and CI.

#### Further details

See https://buildkite.com/docs/pipelines/integrations/plugins/writing#step-5-add-a-test

#### usage

```bash
$ make all
docker compose run lint
TAP version 14
ok 1 - plugin.yml configuration is valid
ok 2 - All the readme config examples for plugin id 'elastic/vault-docker-login' are valid (1 found)
ok 3 - Readme version numbers are up-to-date (v0.5.2)
1..3
# { total: 3, pass: 3 }
# time=549.366ms
docker compose run shellcheck
docker compose \
	  run --rm \
	  	-v /Users/vmartinez/workspaces/elastic/vault-docker-login-buildkite-plugin:/app \
	  	tests
pre-command.bats
 ✓ Clean login execution with kv
pre-exit.bats
 ✓ Clean logout execution
 ✓ No cli available (default case)
 ✓ Skopeo
 ✓ When disable logout is turned on

5 tests, 0 failures
```

v1v committed:
feat: support logout in the pre-exit

Support local development

Add UTs

v1v committed:
Fix tests

v1v committed:
increase test coverage

v1v committed:
bump

v1v commented on Makefile:
```diff
@@ -0,0 +1,20 @@
+clean:
+	-docker compose \
+		rm --force --stop
+	-rm -f reports/* build/* tmp/*
```
```suggestion
```

v1v committed:
Update Makefile

Signed-off-by: Victor Martinez <victormartinezrubio@gmail.com>

esenmarti commented on Makefile:
```diff
@@ -0,0 +1,19 @@
+test:
+	-docker compose \
+	  run --rm \
+	  	-v ${PWD}:/app \
```
```suggestion
	  	-v "${PWD}:/app" \
```

esenmarti commented on Makefile:
```diff
@@ -0,0 +1,19 @@
+
+clean:
+	-docker compose \
+		rm --force --stop
```
```suggestion
		rm --force --stop
		
```

esenmarti commented on README.md:
```diff
@@ -2,6 +2,13 @@
+| Name             | Description                                               | Required | Default        |
+|------------------|-----------------------------------------------------------|----------|----------------|
+| `secret_path`    | The Vault secret path with the user/pass/registry fields. | `true`   | ``             |
+| `disable_logout` | Whether to logout after running the command.              | `false`  | `false`        |
```
nit: Maybe we could add this field to the example below?

esenmarti commented on docker-compose.yml:
```diff
@@ -0,0 +1,19 @@
+    command: -x hooks/pre-command hooks/pre-exit
+    working_dir: /plugin
+    volumes:
+      - ".:/plugin"
```
```suggestion
      - ".:/plugin"
      
```

esenmarti commented on hooks/pre-exit:
```diff
@@ -0,0 +1,36 @@
+  crane auth logout "$REGISTRY_HOSTNAME"
+
+elif cosign version >/dev/null 2>&1; then
+  echo 'WARNING: not logging out supported for cosign cli'
```
```suggestion
  echo 'WARNING: logging out not supported for cosign cli'
```

esenmarti commented on tests/pre-command.bats:
```diff
@@ -0,0 +1,26 @@
+
+setup () {
+  load "${BATS_PLUGIN_PATH}/load.bash"
+  # export DOCKER_STUB_DEBUG=/dev/tty
+  # export VAULT_STUB_DEBUG=/dev/tty
```
Is that used for testing? Maybe add a comment?

v1v replied:
```suggestion
  # NOTE: If you need to debug the docker and vault command output,
  #             you can uncomment the code below:
  # export DOCKER_STUB_DEBUG=/dev/tty
  # export VAULT_STUB_DEBUG=/dev/tty
```

esenmarti reviewed (commented):
some nits

v1v committed:
suggestions

Co-authored-by: Ellie <4158750+esenmarti@users.noreply.github.com>
Signed-off-by: Victor Martinez <victormartinezrubio@gmail.com>

v1v committed:
add example with all the  parameters

esenmarti reviewed (commented):
Also, it looks like the buildkite checks might be failing? I triggered a run (as described [here](https://www.jvt.me/posts/2024/07/16/buildkite-trigger-fork/)): https://buildkite.com/elastic/vault-docker-login-buildkite-plugin/builds/53

v1v commented:
> Also, it looks like the buildkite checks might be failing? I triggered a run (as described [here](https://www.jvt.me/posts/2024/07/16/buildkite-trigger-fork/)): [buildkite.com/elastic/vault-docker-login-buildkite-plugin/builds/53](https://buildkite.com/elastic/vault-docker-login-buildkite-plugin/builds/53)

I cannot reproduce what the CI does easily :/ I'll make sure I can run the same in the Makefile. Interestingly the `make all` does run everything but the pre-commit thingy

v1v committed:
fix

v1v committed:
make pre-commit target

v1v committed:
fix

v1v committed:
chmod

v1v committed:
exclude bats

v1v commented:
I introduced `make pre-commit` and removed the `bats` files from the `shfmt` hook - they are not executable and follow the `BATS` test framework.

Unfortunately I cannot run the CI

jamietanna commented:
Triggered another CI run for ya!

jamietanna commented on .buildkite/pipeline.yml:
```diff
@@ -1,4 +1,8 @@
 steps:
+  - label: ":shell: Tests"
+    plugins:
+      - plugin-tester#v4.1.1: ~
```
https://github.com/buildkite-plugins/plugin-tester-buildkite-plugin/tags
```suggestion
      - plugin-tester#v1.1.1: ~
```

jamietanna committed:
correct version of `plugin-tester`

Signed-off-by: Jamie Tanna <github@jamietanna.co.uk>

v1v committed:
use makefile in the BK pipeline

to help with using the same mechanism locally than in the CI

use GPC with docker otherwise things won't work in container agents

v1v committed:
Merge branch 'main' into feature/support-docker-logout-pre-exit

esenmarti commented on .buildkite/pipeline.yml:
```diff
@@ -1,12 +1,23 @@
+  - label: ":shell: Shellcheck"
+    command: "make shellcheck"
+    agents:
+      provider: "gcp"
+      image: family/core-ubuntu-2204
```
Could we use this ([example](https://github.com/elastic/ci/blob/a036c4d6eeaaf094f5c23d23ba9233ec922622a7/.buildkite/pipeline.yml#L194)) and run `- set -o pipefail ; find . -type f -name '*.sh' -print0 | xargs -0 shellcheck -S error -x -s bash` (or simlar)? Or do we prefer to be consistent and run the same as the make targets?

v1v replied:
I prefer to use Docker instead of reliying the shellcheck tool is installed locally

esenmarti commented on Makefile:
```diff
@@ -0,0 +1,23 @@
+	-docker compose run lint
+
+pre-commit:
+	@echo "Running pre-commit"
```
Should we also add an "echo" in the rest of the targets?

v1v replied:
It's a leftover

v1v replied:
```suggestion
```

esenmarti commented on hooks/pre-exit:
```diff
@@ -0,0 +1,36 @@
+
+elif podman version >/dev/null 2>&1; then
+  echo "Logging out to $REGISTRY_HOSTNAME with podman cli"
+  podman logoutr "$REGISTRY_HOSTNAME"
```
```suggestion
  podman logout "$REGISTRY_HOSTNAME"
```

v1v committed:
Apply suggestions from code review

Co-authored-by: Ellie <4158750+esenmarti@users.noreply.github.com>
Signed-off-by: Victor Martinez <victormartinezrubio@gmail.com>

esenmarti reviewed (approved):
I haven't tested it myself but looks good. Thanks!
